HW1/P4/P4_bfs.py

```diff
@@ -0,0 +1,114 @@
+import findspark
+from functools import partial
+
+findspark.find()
+findspark.init('/usr/local/opt/apache-spark/libexec')
+import pyspark
+
+#########################################################################
+# BFS function
+# Takes in a graph RDD and a character name, update the graph iteratively
+# using Breadth-First Search algorithm. This is the last version that uses 
+# Spark Accumulators and rdd.filters() for optimization
+#########################################################################
+
+def reduce_nodes(nodes_list):
+    nodes_list = list(nodes_list)
+    return sorted(nodes_list, key = lambda x: x[1])[0]
+
+#########################################################################
+# 
+# The update_graph is performed on each node in the filtered range (unused)
+# during each iteration. 
+#
+#########################################################################
+
+def update_graph(node, mc_graph, accum):
+    # extract values from node for clarity
+    node = list((node[0], list(node[1])))
+    character_name = node[0]
+    adj_list = node[1][0]
+    dist = node[1][1]
+    is_used = node[1][2]
+
+    # mc_graph = mc_graph.value
+    result = []
+
+    #if the node has been touched (dist < inf), use it to touch its adjacent nodes
+    if dist < float("inf"):
+        for i in adj_list:
+            #Get the adjacent node
+            new_node = [i, list(mc_graph[i])]
+            # Set distance if it is current min and append it to result list
+            if new_node[1][1] > (dist + 1):
+                new_node[1][1] = (dist + 1)
+                result.append(new_node)
+        # Set the is_used state to True since it has already touch its adj_list
+        node[1][2] = True
+        accum.add(1)
+    # Append the node itself as an element of result
+    result.append(node)
+    return result
+
+
+def BFS(MC_Graph, init_node, sc):
+
+    # Transform MC_Graph (character_name, [adj_list]) to MC_Graph_bfs to 
+    # perform BFS search, MC_Graph_bfs is in the following form:
+    # (character_name, ([adj_list], distance, is_used))
+    #
+    # is_used:Bool value, stores whether the node has been use to activate 
+    # other nodes for distance calculation, used in RDD.filter()
+    #
+    # the distance for every node, except start node, is set to infinite as 
+    # initial state
+
+    MC_Graph_bfs = MC_Graph.map(lambda (x, y): (x, (y, 0, False)) 
+                                if (x == init_node) 
+                                else (x, (y, float("inf"), False)))
+
+    print MC_Graph_bfs.filter(lambda (x, y): x == init_node).collect()
+    # Convert MC_Graph_bfs to adictionary and broadcast it every worker
+    # This is used for lookup during each update
+    tmp = dict(MC_Graph_bfs.collect())
```

Comment:
I think you can comment this out because you didn't use this dictionary any more.